underfs/s3a/src/test/java/alluxio/underfs/s3a/S3AUnderFileSystemMockServerTest.java

```diff
@@ -0,0 +1,135 @@
+import java.util.concurrent.Executors;
+
+/**
+ * Unit tests for the {@link S3AUnderFileSystem} using an s3 mock server.
+ */
+public class S3AUnderFileSystemMockServerTest {
+  private static final InstancedConfiguration CONF = Configuration.copyGlobal();
+
+  private static final String TEST_BUCKET = "test-bucket";
+  private static final String TEST_FILE = "test_file";
+  private static final AlluxioURI TEST_FILE_URI = new AlluxioURI("s3://test-bucket/test_file");
+  private static final String TEST_CONTENT = "test_content";
+
+  private S3AUnderFileSystem mS3UnderFileSystem;
+  private AmazonS3 mClient;
+
+  private S3Mock mS3MockServer;
+
+  @Rule
+  public final ExpectedException mThrown = ExpectedException.none();
+
+  @Before
+  public void before() throws AmazonClientException {
+    mS3MockServer = new S3Mock.Builder().withPort(8001).withInMemoryBackend().build();
+    mS3MockServer.start();
+
+    AwsClientBuilder.EndpointConfiguration
+        endpoint = new AwsClientBuilder.EndpointConfiguration(
+        "http://localhost:8001", "us-west-2");
+    mClient = AmazonS3ClientBuilder
+        .standard()
+        .withPathStyleAccessEnabled(true)
+        .withEndpointConfiguration(endpoint)
+        .withCredentials(new DefaultAWSCredentialsProviderChain())
+        .build();
+    mClient.createBucket(TEST_BUCKET);
+    mS3UnderFileSystem =
+        new S3AUnderFileSystem(new AlluxioURI("s3://" + TEST_BUCKET), mClient, TEST_BUCKET,
+            Executors.newSingleThreadExecutor(), new TransferManager(),
+            UnderFileSystemConfiguration.defaults(CONF), false);
+  }
+
+  @After
+  public void after() {
+    mClient = null;
+    try {
+      if (mS3MockServer != null) {
+        mS3MockServer.shutdown();
+      }
+    } finally {
+      mS3MockServer = null;
+    }
+  }
+
+  @Test
+  public void read() throws IOException {
+    mClient.putObject(TEST_BUCKET, TEST_FILE, TEST_CONTENT);
+
+    InputStream is =
+        mS3UnderFileSystem.open(TEST_FILE_URI.getPath());
+    assertEquals(TEST_CONTENT, IOUtils.toString(is, StandardCharsets.UTF_8));
+  }
+
+  @Test
+  public void listRecursive() throws IOException {
+    mClient.putObject(TEST_BUCKET, "d1/d1/f1", TEST_CONTENT);
+    mClient.putObject(TEST_BUCKET, "d1/d1/f2", TEST_CONTENT);
+    mClient.putObject(TEST_BUCKET, "d1/d2/f1", TEST_CONTENT);
+    mClient.putObject(TEST_BUCKET, "d2/d1/f1", TEST_CONTENT);
+    mClient.putObject(TEST_BUCKET, "d3/", "");
+    mClient.putObject(TEST_BUCKET, "f1", TEST_CONTENT);
+    mClient.putObject(TEST_BUCKET, "f2", TEST_CONTENT);
+
+    UfsStatus[] ufsStatuses = mS3UnderFileSystem.listStatus(
+        "/", ListOptions.defaults().setRecursive(true));
+
+    /*
+      Objects:
+       d1/
+       d1/d1/
+       d1/d1/f1
+       d1/d1/f2
+       d1/d2/
+       d1/d2/f1
+       d2/
+       d2/d1/
+       d2/d1/f1
+       d3/
+       f1
+       f2
+     */
+    assertNotNull(ufsStatuses);
+    assertEquals(12, ufsStatuses.length);
```

Comment:
Do we care about the content and order of files?

Reply:
we do. the goal of this PR is to do some scaffolding and verify the feasibility of the s3 mock server, instead of writing real unit tests to test cover the existing code base. We will refine and add more tests in the future.

Reply:
ok, no problem. good for now.